widget/completionentry.go

```diff
@@ -0,0 +1,203 @@
+	cnv := fyne.CurrentApp().Driver().CanvasForObject(c)
+
+	if c.itemHeight == 0 {
+		// set item height to cache
+		c.itemHeight = c.navigableList.CreateItem().MinSize().Height
+	}
+
+	listheight := float32(len(c.Options))*(c.itemHeight+2*theme.Padding()+theme.SeparatorThicknessSize()) + 2*theme.Padding()
+	canvasSize := cnv.Size()
+	entrySize := c.Size()
+	if canvasSize.Height > listheight {
+		return fyne.NewSize(entrySize.Width, listheight)
+	}
+
+	return fyne.NewSize(
+		entrySize.Width,
+		canvasSize.Height-c.Position().Y-entrySize.Height-theme.InputBorderSize()-theme.Padding())
+}
+
+// Move changes the relative position of the select entry.
+//
+// Implements: fyne.Widget
+func (c *CompletionEntry) Move(pos fyne.Position) {
+	c.Entry.Move(pos)
+	if c.popupMenu != nil {
+		c.popupMenu.Resize(c.maxSize())
+		c.popupMenu.Move(c.popUpPos())
+	}
+}
+
+// Prevent the menu to open when the user validate value from the menu.
+func (c *CompletionEntry) setTextFromMenu(s string) {
+	c.pause = true
+	c.Entry.SetText(s)
+	c.Entry.CursorColumn = len([]rune(s))
+	c.Entry.Refresh()
+	c.pause = false
+	c.popupMenu.Hide()
+}
+
+type navigableList struct {
+	widget.List
+	entry           *widget.Entry
+	selected        int
+	setTextFromMenu func(string)
+	hide            func()
+	navigating      bool
+	items           []string
+}
+
+func newNavigableList(items []string, entry *widget.Entry, setTextFromMenu func(string), hide func()) *navigableList {
+	n := &navigableList{
+		entry:           entry,
+		selected:        -1,
+		setTextFromMenu: setTextFromMenu,
+		hide:            hide,
+		items:           items,
+	}
+
+	n.List = widget.List{
+		Length: func() int {
+			return len(n.items)
+		},
+		CreateItem: func() fyne.CanvasObject {
+			return widget.NewLabel("")
+		},
+		UpdateItem: func(i widget.ListItemID, o fyne.CanvasObject) {
+			o.(*widget.Label).SetText(n.items[i])
+		},
+		OnSelected: func(id widget.ListItemID) {
+			if !n.navigating && id > -1 {
+				setTextFromMenu(n.items[id])
+			}
+			n.navigating = false
+		},
+	}
+	n.ExtendBaseWidget(n)
+	return n
+}
+
+func (n *navigableList) SetOptions(items []string) {
+	n.Unselect(n.selected)
+	n.items = items
+	n.Refresh()
+	n.selected = -1
+}
+
+func (n *navigableList) TypedKey(event *fyne.KeyEvent) {
+	switch event.Name {
+	case fyne.KeyDown:
+		if n.selected < len(n.items)-1 {
+			n.selected++
+		} else {
+			n.selected = 0
+		}
+		n.navigating = true
+		n.Select(n.selected)
+
+	case fyne.KeyUp:
+		if n.selected > 0 {
+			n.selected--
+		} else {
+			n.selected = len(n.items) - 1
+		}
+		n.navigating = true
+		n.Select(n.selected)
+	case fyne.KeyReturn, fyne.KeyEnter:
+		if n.selected == -1 { // so the user want to submit the entry
+			n.hide()
+			n.entry.TypedKey(event)
+		} else {
+			n.navigating = false
+			n.OnSelected(n.selected)
+		}
+	case fyne.KeyEscape:
+		n.hide()
+	default:
+		n.entry.TypedKey(event)
+
+	}
+}
+func (n *navigableList) TypedRune(r rune) {
```

Comment:
Can you leave a space between methods please?

Reply:
For TypedRune method it's OK to add a new line, but for the 2 latest, adding a new line is weird IMHO and will remove the alignment of brackets:

``` go
func (n *navigableList) FocusGained() {}
func (n *navigableList) FocusLost()   {}

// after goimports (automatic in my editor)
func (n *navigableList) FocusGained() {}

func (n *navigableList) FocusLost() {}

```

Reply:
I did changes, tell me if there are other things to do ;)